File: backend/src/main/java/com/company/projectmanagement/service/ActivityService.java
// ================= ActivityService.java =================
package com.company.projectmanagement.service;

import com.company.projectmanagement.dto.ActivityHistoryDto;
import com.company.projectmanagement.entity.ActivityHistory;
import com.company.projectmanagement.repository.ActivityHistoryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class ActivityService {

    private final ActivityHistoryRepository repository;

    /* ================= GET TASK HISTORY ================= */
    public List<ActivityHistoryDto> getTaskHistory(Long taskId) {
        return repository.findByTaskIdOrderByPerformedAtDesc(taskId)
                .stream()
                .map(this::mapToDto)
                .collect(Collectors.toList());
    }

    /* ================= CREATE ACTIVITY ================= */
    public ActivityHistoryDto createActivity(ActivityHistoryDto dto) {

        ActivityHistory entity = ActivityHistory.builder()
                .taskId(dto.getTaskId())
                .activityType(dto.getActivityType())
                .oldValue(dto.getOldValue())
                .newValue(dto.getNewValue())
                .performedBy(dto.getPerformedBy())
                .build();

        ActivityHistory saved = repository.save(entity);
        return mapToDto(saved);
    }

    /* ================= MAPPER ================= */
    private ActivityHistoryDto mapToDto(ActivityHistory a) {
        return ActivityHistoryDto.builder()
                .id(a.getId())
                .taskId(a.getTaskId())
                .activityType(a.getActivityType())
                .oldValue(a.getOldValue())
                .newValue(a.getNewValue())
                .performedBy(a.getPerformedBy())
                .performedAt(a.getPerformedAt())
                .build();
    }
}

File: frontend/src/services/activityService.js
import API from "./api";

/* ================= GET TASK HISTORY ================= */
export const getTaskHistory = async (taskId) => {
  try {
    const response = await API.get(`/activity/task/${taskId}`);
    return response.data;
  } catch (error) {
    console.error(`Error fetching activity history for task ${taskId}:`, error);
    throw error;
  }
};

/* ================= CREATE ACTIVITY ================= */
export const createActivity = async (activity) => {
  try {
    const response = await API.post("/activity", activity);
    return response.data;
  } catch (error) {
    console.error("Error creating activity:", error);
    throw error;
  }
};

export const getActivity = getTaskHistory;

